torch/csrc/autograd/variable.cpp

```diff
@@ -170,7 +170,8 @@ void Variable::Impl::set_data(const at::Tensor &new_data) {
   device_opt_ = new_data.device();
   type_id_ = new_data.dispatch_type().type_id();
 
-  auto new_data_impl_copy = new_data.getIntrusivePtr()->shallow_copy_and_detach();
+  auto new_data_impl_copy = new_data.getIntrusivePtr()->shallow_copy_and_detach(
```

Comment:
wait, so we never actually share the version counter?

Perhaps a better API, then, is to just pass the version counter you want to shallow_copy_and_detach?

Reply:
We will have three cases for how we handle the version counter in https://github.com/pytorch/pytorch/pull/17072:
1. Share the version counter
2. Create a new version counter
3. Keep the destination TensorImpl's original version counter (only applicable to `shallow_copy_from()`, for usage in `Variable::set_data()`)

I think allowing the user to pass the version counter they want to `shallow_copy_and_detach()` and `shallow_copy_from()` would be a better API, as it would allow the most flexibility.

Reply:
that sounds good to me.